### PR TITLE
ENH: Improve to_string typing

### PR DESCRIPTION
**Pull request by bashtage**
- [x] Closes #xxxx (Replace xxxx with the Github issue number)
- [x] Tests added: Please use `assert_type()` to assert the type of any return value

### CODE REVIEW

File: pandas-stubs/_typing.pyi [hunk 1]
```diff
@@ -186,7 +187,10 @@ CompressionDict = dict[str, Any]
 CompressionOptions = Optional[
     Union[Literal["infer", "gzip", "bz2", "zip", "xz", "zstd"], CompressionDict]
 ]
-
+FormattersType = Union[
+    list[Callable], tuple[Callable, ...], Mapping[Union[str, int], Callable]
+]
+FloatFormatType = str | Callable | matplotlib.ticker.EngFormatter
```
Comment (Dr-Irv) [inline pandas-stubs/_typing.pyi:193]: pandas uses `pandas.io.formats.format.EngFormatter`, so import that, and you remove the `matplotlib` issue.

File: pandas-stubs/core/frame.pyi [hunk 2]
```diff
@@ -1962,48 +1964,49 @@ class DataFrame(NDFrame, OpsMixin):
     @overload
     def to_string(
         self,
-        buf: FilePathOrBuffer | None,
+        buf: FilePath | WriteBuffer[str],
         columns: Sequence[_str] | None = ...,
```
[the remark this answers is not in the document]
Comment (bashtage) [reply pandas-stubs/core/frame.pyi]: Fixed.

File: pandas-stubs/_typing.pyi [hunk 3]
```diff
@@ -23,6 +23,7 @@ from typing import (
     Union,
 )
 
+import matplotlib.ticker
```
[the remark this answers is not in the document]
Comment (bashtage) [reply pandas-stubs/_typing.pyi]: Fixed.

File: pandas-stubs/_typing.pyi (hunk 1 above)
Comment (bashtage) [reply pandas-stubs/_typing.pyi]: Fixed.

File: pandas-stubs/core/frame.pyi [hunk 4]
```diff
@@ -2022,7 +2022,7 @@ class DataFrame(NDFrame, OpsMixin):
     def to_string(
         self,
         buf: None = ...,
-        columns: Sequence[_str] | None = ...,
+        columns: list[HashableT] | Index | None = ...,
```
Comment (Dr-Irv) [inline pandas-stubs/core/frame.pyi:2025]: Shouldn't `Series` be here as well?

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
One small thing I caught.

File: pandas-stubs/core/frame.pyi (hunk 4 above)
Comment (bashtage) [reply pandas-stubs/core/frame.pyi]: Added

**Comment by bashtage** [comment]
Fixed the small things and merged main.

**Review by Dr-Irv — APPROVED** [review]
thanks @bashtage

**Pull request merged**